File: frontend/src/api/portal.ts
import portalClient from './portalClient';

export interface PortalTicket {
  id: string;
  subject: string;
  status: string;
  priority: string;
  requester_email: string | null;
  created_at: string;
  updated_at: string;
}

export interface PortalMessage {
  id: string;
  sender_type: string;
  body_text: string | null;
  body_html: string | null;
  created_at: string;
}

export interface PortalArticle {
  id: string;
  title: string;
  slug: string;
  body_html: string;
  category_id: string | null;
  published_at: string | null;
}

export const portalAuthApi = {
  register: (data: { email: string; password: string; display_name?: string }) =>
    portalClient.post<{ access_token: string }>('/portal/auth/register', data).then(r => r.data),
  login: (data: { email: string; password: string }) =>
    portalClient.post<{ access_token: string }>('/portal/auth/login', data).then(r => r.data),
  me: () => portalClient.get('/portal/auth/me').then(r => r.data),
};

export const portalTicketApi = {
  list: (params?: { page?: number; page_size?: number; status?: string }) =>
    portalClient.get<{ items: PortalTicket[]; total: number; page: number; page_size: number }>('/portal/tickets', { params }).then(r => r.data),
  create: (data: { subject: string; body: string }) =>
    portalClient.post<PortalTicket>('/portal/tickets', data).then(r => r.data),
  createGuest: (data: { email: string; subject: string; body: string }) =>
    portalClient.post<PortalTicket>('/portal/tickets/guest', data).then(r => r.data),
  get: (id: string) => portalClient.get<PortalTicket>(`/portal/tickets/${id}`).then(r => r.data),
  getMessages: (id: string) => portalClient.get<PortalMessage[]>(`/portal/tickets/${id}/messages`).then(r => r.data),
  reply: (id: string, body_text: string) =>
    portalClient.post<PortalMessage>(`/portal/tickets/${id}/reply`, { body_text }).then(r => r.data),
};

export const portalKbApi = {
  listArticles: (params?: { page?: number; page_size?: number; category_id?: string }) =>
    portalClient.get<{ items: PortalArticle[]; total: number; page: number; page_size: number }>('/portal/kb/articles', { params }).then(r => r.data),
  getArticle: (slug: string) => portalClient.get<PortalArticle>(`/portal/kb/articles/${slug}`).then(r => r.data),
  listCategories: () => portalClient.get('/portal/kb/categories').then(r => r.data),
  search: (data: { query: string; page?: number; page_size?: number }) =>
    portalClient.post('/portal/kb/search', data).then(r => r.data),
};

File: backend/app/schemas/portal.py
from pydantic import BaseModel
from typing import Optional
from datetime import datetime


class CustomerRegister(BaseModel):
    email: str
    password: str
    display_name: Optional[str] = None


class CustomerLogin(BaseModel):
    email: str
    password: str


class CustomerTokenResponse(BaseModel):
    access_token: str


class CustomerOut(BaseModel):
    id: str
    email: str
    display_name: Optional[str] = None
    is_registered: bool
    created_at: datetime

    class Config:
        from_attributes = True


class PortalTicketCreate(BaseModel):
    subject: str
    body: str
    requester_email: Optional[str] = None


class PortalTicketOut(BaseModel):
    id: str
    subject: str
    status: str
    priority: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PortalTicketDetailOut(PortalTicketOut):
    requester_email: Optional[str] = None


class PortalTicketListResponse(BaseModel):
    items: list[PortalTicketOut]
    total: int
    page: int
    page_size: int


class PortalMessageCreate(BaseModel):
    body_text: str


class PortalMessageOut(BaseModel):
    id: str
    sender_type: str
    body_text: Optional[str] = None
    body_html: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class PortalArticleOut(BaseModel):
    id: str
    title: str
    slug: str
    body_html: str
    category_id: Optional[str] = None
    published_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class PortalArticleListResponse(BaseModel):
    items: list[PortalArticleOut]
    total: int
    page: int
    page_size: int
